minidump/src/minidump.rs

```diff
@@ -926,15 +928,21 @@ impl Module for MinidumpModule {
             _ => None,
         }
     }
-    fn debug_identifier(&self) -> Option<Cow<'_, str>> {
+    fn debug_identifier(&self) -> Option<DebugId> {
         match self.codeview_info {
             Some(CodeView::Pdb70(ref raw)) => {
-                let id = format!("{:#}{:x}", raw.signature, raw.age,);
-                Some(Cow::Owned(id))
+                let uuid = Uuid::from_fields(
+                    raw.signature.data1,
+                    raw.signature.data2,
+                    raw.signature.data3,
+                    &raw.signature.data4,
+                )
+                .ok()?;
+                Some(DebugId::from_parts(uuid, raw.age))
             }
             Some(CodeView::Pdb20(ref raw)) => {
-                let id = format!("{:08X}{:x}", raw.signature, raw.age);
-                Some(Cow::Owned(id))
+                let uuid = Uuid::from_fields(raw.signature, 0, 0, &[0, 0, 0, 0]).ok()?;
+                Some(DebugId::from_parts(uuid, raw.age))
```

Comment:
This `Pdb20` branch does not have any test coverage.  I think this is the right way to convert it but I'm not sure.  The previous generated string was also not a valid breakpad debugid because it was too short.

Reply:
This isn't correct, and Breakpad isn't the source of truth on this matter. However, this is also an obsolete format, I think the last compiler to produce it was Visual C++ 6.0. I just shared this link (which is in the crate docs somewhere but now 404s) today, it mentions what's going on there:
http://web.archive.org/web/20210227224734/https://www.debuginfo.com/articles/debuginfomatch.html

Reply:
From that link I think it only says the signature is a timestamp in seconds since the epoch.  The comment at https://github.com/getsentry/breakpad/blob/63c90f00bf64a1b4a8883e50bbe8217048ed2a85/src/common/windows/pdb_source_line_writer.cc#L1217-L1229 seems to suggest that in this case the GUID conversion happens with 0-filled.  Technically this is an invalid UUID though, since the UUID variant flag (1st 3 bits of byte 33) is set to the legacy "variant 0" but then technically it is impossible to create a valid UUID out of this.

At the end of the day we're doing something to get some kind of DebugId-shaped identifier for which te symbol provider would have to do the same transformation and I think this is the same as breakpad was suggesting it might do in this case and probably good enough until/if someone ever encounters such a debugid?  So I think after learning more about this this is still a reasonable solution here.

Reply:
That's my point—it's *not* a UUID at all! Here, I realized that an old build of Firefox would have PDB 2.0 debug info, so if you download [the Firefox 1.0 win32 build](https://ftp.mozilla.org/pub/firefox/releases/1.0/win32/en-US/Firefox%201.0.zip), you can see what `CV_INFO_PDB20` looks like. I poked at the `firefox.exe`from there in WinDBG:

```
0:000> !sym noisy
noisy mode - symbol prompts on
0:000> .reload -f
.
DBGHELP: C:\Windows\system32\firefox.exe - file not found
DBGENG:  C:\Users\ted\Desktop\firefox.exe - Mapped image memory
SYMSRV:  BYINDEX: 0x3
         C:\ProgramData\Dbg\sym
         firefox.pdb
         418E89C31
SYMSRV:  UNC: C:\ProgramData\Dbg\sym\firefox.pdb\418E89C31\firefox.pdb - path not found
SYMSRV:  UNC: C:\ProgramData\Dbg\sym\firefox.pdb\418E89C31\firefox.pd_ - path not found
SYMSRV:  UNC: C:\ProgramData\Dbg\sym\firefox.pdb\418E89C31\file.ptr - path not found
SYMSRV:  RESULT: 0x80070003
SYMSRV:  BYINDEX: 0x4
         C:\ProgramData\Dbg\sym*https://msdl.microsoft.com/download/symbols
         firefox.pdb
         418E89C31
SYMSRV:  UNC: C:\ProgramData\Dbg\sym\firefox.pdb\418E89C31\firefox.pdb - path not found
SYMSRV:  UNC: C:\ProgramData\Dbg\sym\firefox.pdb\418E89C31\firefox.pd_ - path not found
SYMSRV:  UNC: C:\ProgramData\Dbg\sym\firefox.pdb\418E89C31\file.ptr - path not found
SYMSRV:  HTTPGET: /download/symbols/firefox.pdb/418E89C31/firefox.pdb
SYMSRV:  HttpQueryInfo: 80190194 - HTTP_STATUS_NOT_FOUND
```

You can see in the output that the debug id is `418E89C31`—that's `signature == 0x418E89C3` and `age == 1`. The current implementation should produce the same result here, but your replacement will not.

I'm not really worried about supporting these in practice, as they're obsolete, but it doesn't make sense to take code that is correct and break it.

Reply:
I get that this is not a UUID, to look it up though we have to somehow create a DebugId from it however and filling the remainder of the UUID bytes with zeroes is exactly the same approach as breakpad does at https://github.com/getsentry/breakpad/blob/63c90f00bf64a1b4a8883e50bbe8217048ed2a85/src/common/windows/pdb_source_line_writer.cc#L1217-L1229 which makes me think this is reasonable.  It also is simply what we can declare is done in order to map this old format, but if I understand you don't like this solution.

That leaves only two other options:
- The debugid crate, in its current form, is not suitable to represent all Debug Information File identifiers so can not be adopted.
- The minidump crate simply does not care about Pdb20 anymore and will not return DIF identifiers for them.

While it is perfectly possible to make a debugid::DebugId that has a specific format for Pdb20 I do think it the solution here (and in breakpad IIUC) to simply map it to an invalid UUID is more practical given how rare this is.  We could also compromise and make this more explicit by moving this logic into the debugid crate by adding a `DebugId::from_pdb20_signature` constructor and describing what is being done.

Reply:
> A more thorough solution would be to extend the debugid crate to handle these obsolete formats, but that doesn't seem particularly worthwhile given that they're unlikely to be encountered in the wild nowadays.

This is the option I"m exploring in https://github.com/getsentry/rust-debugid/pull/10 though that is not the right implementation at all currently.  I have something that's actually identical memory-layout wise locally that i'll finish tomorrow.

Reply:
Thanks! I'm amenable to either path. I appreciate you doing the work here, I know this line of inquiry feels very nit-picky but I have run into these sorts of weird edge cases enough in my life to be wary of leaving them for someone else to hit.